repair/row_level.cc

```diff
@@ -1870,37 +1882,58 @@ class repair_meta {
     future<> put_row_diff_with_rpc_stream(
             repair_hash_set set_diff,
             needs_all_rows_t needs_all_rows,
-            gms::inet_address remote_node, unsigned node_idx) {
-        if (!set_diff.empty()) {
-            if (remote_node == _myip) {
-                return make_ready_future<>();
-            }
-            size_t sz = set_diff.size();
-            return get_row_diff(std::move(set_diff), needs_all_rows).then([this, remote_node, node_idx, sz] (std::list<repair_row> row_diff) {
-                if (row_diff.size() != sz) {
-                    rlogger.warn("Hash conflict detected, keyspace={}, table={}, range={}, row_diff.size={}, set_diff.size={}. It is recommended to compact the table and rerun repair for the range.",
-                            _schema->ks_name(), _schema->cf_name(), _range, row_diff.size(), sz);
+            gms::inet_address remote_node, unsigned node_idx,
+            const locator::effective_replication_map& erm, bool small_table_optimization) {
+        if (set_diff.empty()) {
+            co_return;
+        }
+        if (remote_node == _myip) {
+            co_return;
+        }
+        size_t sz = set_diff.size();
+
+        std::list<repair_row> row_diff = co_await get_row_diff(std::move(set_diff), needs_all_rows);
+        if (row_diff.size() != sz) {
+            rlogger.warn("Hash conflict detected, keyspace={}, table={}, range={}, row_diff.size={}, set_diff.size={}. It is recommended to compact the table and rerun repair for the range.",
+                    _schema->ks_name(), _schema->cf_name(), _range, row_diff.size(), sz);
+        }
+        if (small_table_optimization) {
+            auto& strat = erm.get_replication_strategy();
+            auto& tm = erm.get_token_metadata();
+            std::list<repair_row> tmp;
+            for (auto& row : row_diff) {
+                repair_row r = std::move(row);
+                const auto& dk = r.get_dk_with_hash()->dk;
+                auto eps = co_await strat.calculate_natural_endpoints(dk.token(), tm);
+                if (eps.contains(remote_node)) {
+                    tmp.push_back(std::move(r));
+                } else {
+                    rlogger.trace("master: put : ignore row, token={}", dk.token());
```

Comment:
If the sender never send partitions that do not belong to the receiver. There is no need for the receiver to check. It is a big win to check on the sender 1) avoids sending something will be dropped by the receiver for no reason 2) this will work even on a mixed cluster which runs the old version of scylla 3) The whole logic runs on the master node only.